File: opensource/engine-persistence/src/main/java/com/sos/scheduler/engine/persistence/entities/JobEntity.java
package com.sos.scheduler.engine.persistence.entities;

import com.sos.scheduler.engine.data.scheduler.ClusterMemberId;
import com.sos.scheduler.engine.data.scheduler.SchedulerId;

import javax.annotation.Nullable;
import javax.persistence.*;
import java.io.Serializable;
import java.util.Date;

import static com.google.common.base.Strings.isNullOrEmpty;
import static com.sos.scheduler.engine.persistence.SchedulerDatabases.schedulerIdFromDatabase;
import static javax.persistence.TemporalType.TIMESTAMP;

@Entity
@Table(name="SCHEDULER_JOBS")
@IdClass(JobEntity.PrimaryKey.class)
public class JobEntity {
    @Column(name="spooler_id") @Id private String schedulerId;
    @Column(name="cluster_member_id") @Id private String clusterMemberId;
    @Column(name="path") @Id private String jobPath;
    @Column(name="stopped") private boolean stopped;
    @Column(name="next_start_time") @Temporal(TIMESTAMP) private Date nextStartTime;

    public SchedulerId getSchedulerId() {
        return schedulerIdFromDatabase(schedulerId);
    }

    public ClusterMemberId getClusterMemberId() {
        return new ClusterMemberId(clusterMemberId.equals("-")? "" : clusterMemberId);
    }

    @Nullable public final String getJobPath() {
        return isNullOrEmpty(jobPath)? null : "/" + jobPath;
    }

    public final boolean isStopped() {
        return stopped;
    }

    @Nullable public final Date getNextStartTime() {
        return nextStartTime == null? null : (Date)nextStartTime.clone();
    }

    public static class PrimaryKey implements Serializable {
        @Column(name="spooler_id") private String schedulerId;
        @Column(name="cluster_member_id") private String clusterMemberId;
        @Column(name="path") private String jobPath;
    }
}

File: opensource/engine-persistence/src/main/scala/com/sos/scheduler/engine/persistence/entities/JobEntity.scala
package com.sos.scheduler.engine.persistence.entities

import java.io.Serializable
import java.util.{Date => JavaDate}
import javax.annotation.Nullable
import javax.persistence.TemporalType.TIMESTAMP
import javax.persistence._

/** JPA-Entity für einen Jobzustand.
  * <pre>
  * CREATE TABLE SCHEDULER_JOBS (
  *   SPOOLER_ID varchar(100) NOT NULL,
  *   CLUSTER_MEMBER_ID varchar(100) NOT NULL,
  *   PATH varchar(255) NOT NULL,
  *   STOPPED boolean NOT NULL,
  *   NEXT_START_TIME varchar(24),
  *   PRIMARY KEY (spooler_id, cluster_member_id, path));
  * CREATE UNIQUE INDEX PRIMARY_KEY_B ON SCHEDULER_JOBS (SPOOLER_ID, CLUSTER_MEMBER_ID, PATH);
  * </pre>*/
@Entity
@Table(name = "SCHEDULER_JOBS")
@IdClass(classOf[JobEntity.PrimaryKey])
class JobEntity {

  @Column(name=""""SPOOLER_ID"""", nullable=false) @Id
  private [entities] var schedulerId: String = _

  @Column(name=""""CLUSTER_MEMBER_ID"""", nullable=false) @Id
  private [entities] var clusterMemberId: String = _

  @Column(name=""""PATH"""", nullable=false) @Id
  private [entities] var jobPath: String = _

  @Column(name=""""STOPPED"""" , nullable=false)
  private [entities] var isStopped: Boolean = _

  @Column(name=""""NEXT_START_TIME"""") @Temporal(TIMESTAMP) @Nullable
  private [entities] var nextStartTime: JavaDate = _

  override def toString = "JobEntity"+ Seq(schedulerId, clusterMemberId, jobPath, isStopped, nextStartTime).mkString("(", ",", ")")
}

object JobEntity {
  case class PrimaryKey(var schedulerId: String, var clusterMemberId: String, var jobPath: String) extends Serializable {
    def this() = this(null: String, null: String, null: String)
  }
}
